File: Controller/src/com/psu/capstone/tiutracking/Vector2D.java
package com.psu.capstone.tiutracking;

public class Vector2D {
  public double x, y;
  
  public Vector2D() {
    this(0, 0);
  }
  
  public Vector2D(double x, double y) {
    this.x = x;
    this.y = y;
  }
  
  public Vector2D(Vector2D v) {
    this(v.x, v.y);
  }
  
  public void add(double x, double y) {
    this.x += x;
    this.y += y;
  }
  
  public void add(Vector2D v) {
    add(v.x, v.y);
  }
  
  public void sub(Vector2D v) {
    sub(v.x, v.y);
  }
  
  public void sub(double x, double y) {
    this.x -= x;
    this.y -= y;
  }
  
  public void mult(double x, double y) {
    this.x *= x;
    this.y *= y;
  }
  
  public void mult(Vector2D v) {
    mult(v.x, v.y);
  }
  
  public void mult(double c) {
    mult(c, c);
  }
  
  public void lerp(Vector2D other, double factor) {
	x += (other.x - x) * factor;
	y += (other.y - y) * factor;
  }
  
  public void lerp(Vector2D a, Vector2D b, double factor) {
	x = a.x + (b.x - a.x) * factor;
	y = b.y + (b.y - a.y) * factor;
  }
  
  public double mag() {
    if (x == 0 && y == 0) {
      return 0;
    }
	return Math.sqrt(x * x + y * y);
  }
  
  public double magSquared() {
    return x * x + y * y;
  }
  
  public void set(Vector2D v) {
    x = v.x;
    y = v.y;
  }
  
  public static Vector2D sub(Vector2D a, Vector2D b) {
    return new Vector2D(a.x - b.x, a.y - b.y);
  }
  
  public static Vector2D mult(Vector2D v, double c) {
    return new Vector2D(v.x * c, v.y * c);
  }
}

File: trunk/Software/WebApp/src/Vector2D.js
/**
 *	TVector2D
 *
 *	@author	Man Hoang
 *	@version	1.0
 */
function TVector2D() {
	this.x = 0;
	this.y = 0;
}

TVector2D.prototype = {
	assign: function (other) {
		this.x = other.x;
		this.y = other.y;
	},
	
	multAddSet: function (v1, v2, s) {
		this.x = v1.x + v2.x * s;
		this.y = v1.y + v2.y * s;
	},
	
	multSubSet: function (v1, v2, s) {
		this.x = v1.x - v2.x * s;
		this.y = v1.y - v2.y * s;
	},
	
	equals: function (other, epsilon) {
		return	isZero(this.x - other.x, epsilon) &&
				isZero(this.y - other.y, epsilon);
	},
	
	length: function () {
		return Math.sqrt(this.x*this.x + this.y*this.y);
	},

	squaredLength: function () {
		return this.x*this.x + this.y*this.y;
	},

	distanceTo: function (other) {
		var dx = this.x - other.x;
		var dy = this.y - other.y;
		return Math.sqrt(dx*dx + dy*dy);
	},

	lerp: function (v1, v2, factor) {
		this.x = v1.x + factor * (v2.x - v1.x);
		this.y = v1.y + factor * (v2.y - v1.y);
	},
	
	add: function (other) {
		this.x += other.x;
		this.y += other.y;
	},
	
	sub: function (other) {
		this.x -= other.x;
		this.y -= other.y;
	},
	
	mult: function (c) {
		this.x *= c;
		this.y *= c;
	},
	
	dotProduct: function (other) {
		return this.x * other.x + this.y * other.y;
	},
	
	toString: function () {
		return '(X: ' + this.x.toFixed(1) + ', Y: ' + this.y.toFixed(1) + ')';
	}
};

/* Global constants */

ZeroVector2D = new TVector2D();
